pyiceberg/table/name_mapping.py

```diff
@@ -0,0 +1,203 @@
+            raise ValueError(f"Invalid mapping: ID {field.field_id} is not unique")
+
+        self.result[field.field_id] = field
+
+        return self.result
+
+
+class _IndexByName(NameMappingVisitor[Dict[str, MappedField]]):
+    def mapping(self, nm: NameMapping, field_results: Dict[str, MappedField]) -> Dict[str, MappedField]:
+        return field_results
+
+    def fields(self, struct: List[MappedField], field_results: List[Dict[str, MappedField]]) -> Dict[str, MappedField]:
+        return dict(ChainMap(*field_results))
+
+    def field(self, field: MappedField, field_result: Dict[str, MappedField]) -> Dict[str, MappedField]:
+        result: Dict[str, MappedField] = {
+            f"{field_name}.{key}": result_field for key, result_field in field_result.items() for field_name in field.names
+        }
+
+        for name in field.names:
+            result[name] = field
+
+        return result
+
+
+@singledispatch
+def visit_name_mapping(obj: Union[NameMapping, List[MappedField], MappedField], visitor: NameMappingVisitor[T]) -> T:
+    """Traverse the name mapping in post-order traversal."""
+    raise NotImplementedError(f"Cannot visit non-type: {obj}")
+
+
+@visit_name_mapping.register(NameMapping)
+def _(obj: NameMapping, visitor: NameMappingVisitor[T]) -> T:
+    return visitor.mapping(obj, visit_name_mapping(obj.root, visitor))
+
+
+@visit_name_mapping.register(list)
+def _(fields: List[MappedField], visitor: NameMappingVisitor[T]) -> T:
+    results = [visitor.field(field, visit_name_mapping(field.fields, visitor)) for field in fields]
+    return visitor.fields(fields, results)
+
+
+def load_mapping_from_json(mapping: str) -> NameMapping:
+    return NameMapping.model_validate_json(mapping)
+
+
+class _CreateMapping(SchemaVisitor[List[MappedField]]):
+    def schema(self, schema: Schema, struct_result: List[MappedField]) -> List[MappedField]:
+        return struct_result
+
+    def struct(self, struct: StructType, field_results: List[List[MappedField]]) -> List[MappedField]:
+        return [
+            MappedField(field_id=field.field_id, names=[field.name], fields=result)
+            for field, result in zip(struct.fields, field_results)
+        ]
+
+    def field(self, field: NestedField, field_result: List[MappedField]) -> List[MappedField]:
+        return field_result
+
+    def list(self, list_type: ListType, element_result: List[MappedField]) -> List[MappedField]:
+        return [MappedField(field_id=list_type.element_id, names=["element"], fields=element_result)]
+
+    def map(self, map_type: MapType, key_result: List[MappedField], value_result: List[MappedField]) -> List[MappedField]:
+        return [
+            MappedField(field_id=map_type.key_id, names=["key"], fields=key_result),
+            MappedField(field_id=map_type.value_id, names=["value"], fields=value_result),
+        ]
+
+    def primitive(self, primitive: PrimitiveType) -> List[MappedField]:
+        return []
```

Comment:
Shouldn't this be `None`?

Reply:
Then all the signatures should change into `Optional[List[MappedField]]` which isn't nice. I've commented above: https://github.com/apache/iceberg-python/pull/212#discussion_r1429740418